torch/csrc/jit/passes/concat_opt.cpp

```diff
@@ -510,5 +510,65 @@ void OptimizeConcat(const std::shared_ptr<Graph>& graph) {
   GRAPH_DUMP("After ConcatOpt", graph);
 }
 
+namespace {
+
+class VariadicCatUpdater {
+ public:
+  explicit VariadicCatUpdater(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {}
+
+  void run() {
+    collectCatNodes(graph_->block());
+    for (auto c : cat_nodes_) {
+      replaceWithVariadicCat(c);
+    }
+  }
+
+ private:
+  void collectCatNodes(Block* block) {
+    for (auto node : block->nodes()) {
+      if (node->kind() == aten::cat) {
+        cat_nodes_.push_back(node);
+      }
+      for (Block* block : node->blocks()) {
+        collectCatNodes(block);
+      }
+    }
+  }
+
+  void replaceWithVariadicCat(Node* cat) {
+    if (cat->input(0)->node()->kind() != prim::ListConstruct) {
+      return;
+    }
+    auto list = cat->input(0)->node();
+    // We do not transform cat ops whose list input has > 1 use. This is
+    // because these uses could be modifying the list using ops like
+    // `aten::append`. So, we conservatively assume that any use other than
+    // the one in cat mutates the list.
+    if (list->output()->uses().size() > 1) {
```

Comment:
Just doing a quick review, I think we want this pass to work by testing the `aliasDb. moveBeforeTopologicallyValid`, that way we can get the concat optimization pass you wrote firing with DensetNet even though the list is appended to. What do you think ?

Reply:
As we discussed in the meeting, the plan would be to remove list mutation before doing this transformation. Given that, why do we need to use `aliasDb.moveBeforeTopologicallyValid` here? 

I definitely want to apply the concat optimization pass for DenseNet, but that can be done as soon as we eliminate list mutation.